noobaa_sa/s3_client.py

```diff
@@ -312,3 +308,95 @@ def delete_all_objects_in_bucket(self, bucket_name):
         # TODO: Support buckets with versioning enabled
         log.info(f"Deleting all objects in bucket {bucket_name} via boto3")
         self._boto3_resource.Bucket(bucket_name).objects.all().delete()
+
+    def initiate_multipart_object_upload(self, bucket_name, object_name):
+        """
+        Initiate multipart object to the S3 bucket using boto3
+
+         Args:
+            bucket_name (str): The name of the S3 bucket.
+            object_name(str): The unique name of the S3 object
+
+        Returns:
+            Str: Upload id generated by boto3 client
+        """
+
+        resp = self._boto3_client.create_multipart_upload(
+            Bucket=bucket_name, Key=object_name
+        )
+        return resp['UploadId']
+
+    def initiate_upload_part(
+            self,
+            upload_id,
+            bucket_name,
+            object_name,
+            origin_dir,
+            part_size="10M",
+            ):
+        """
+        Upload multiple parts of the file as object to the bucket using boto3
+
+         Args:
+            upload_id (str): id generated by create_multipart_upload method
+            bucket_name (str): The name of the S3 bucket.
+            object_name(str): The unique name of the S3 object
+            origin_dir (str): A directory where the files are present.
+            part_size (String): Part size
+
+        Returns:
+            List: List contains all part information
+        """
+        file_name = origin_dir + "/" + object_name
+        all_part_info = []
+        part_id = 1
+        new_part_size = parse_size_to_bytes(part_size)
+        with open(file_name, "rb") as f:
+            while True:
+                log.info(f"Uploading 10 MB chunks of the file {file_name}")
+                file_chunk = f.read(new_part_size)
+                if file_chunk == b'':
+                    break
+                part_info = self._boto3_client.upload_part(
+                    Bucket=bucket_name,
+                    Key=object_name,
+                    PartNumber=part_id,
+                    UploadId=upload_id,
+                    Body=file_chunk
+                )
+                all_part_info.append({
+                    'PartNumber': part_id,
+                    'ETag': part_info['ETag']
+                })
+                part_id += 1
+        return all_part_info
```

Comment:
This still lacks modularity:

1. The method uploads all the chunks in a row, so we can't test how doing other operations in between might affect the result.
2. It also forces us to upload all the parts with the same chunk size. One might want to test uploading the parts in random sizes.

I suggest adding an `upload_part` method to S3Client instead, which will basically just call `self._boto3_client.upload_part` here, and then pack the logic from this original method to a separate method (or util funciton?) named `upload_object_in_even_multiparts`, using the 3 atomic operations `initiate_multipart_object_upload`, `upload_part` and `complete_multipart_object_upload`.
 
The above might seem redundant since it basically just wrapping the boto3 client, but it has some benefits:
-  The modularity needed for the examples above
-  We can add error handling and logging to each of these methods
-  With the plan to make this class a boto3 inheritor of an abstract S3Client class in mind, the other inheritors will have different implementations for these simple methods.

Reply:
Fixed